src/conv/invokers/mlir_impl_gemm.cpp

```diff
@@ -0,0 +1,150 @@
+// infer actual layout
+// - For NCHW, sizes = {N, C, H, W}; strides = {C*H*W, H*W, W, 1}
+// - For NHWC, sizes = {N, C, H, W}; strides = {C*H*W, 1, W*C, C}
+auto permuteDimsStrides(const std::vector<size_t>& dims, const std::vector<size_t>& strides)
+{
+    auto sorted_dims    = dims;
+    auto sorted_strides = strides;
+    auto p              = TensorDescriptor::sort_permutation(strides, std::greater<>{});
+    std::transform(p.begin(), p.end(), sorted_dims.begin(), [&](auto i) { return dims[i]; });
+    std::transform(p.begin(), p.end(), sorted_strides.begin(), [&](auto i) { return strides[i]; });
+    return std::make_tuple(sorted_dims, sorted_strides);
+};
+
+void permuteDimStridesAllDir(const conv::ProblemDescription& conv_problem,
+                             std::vector<size_t>& in_dims,
+                             std::vector<size_t>& in_strides,
+                             std::vector<size_t>& weights_dims,
+                             std::vector<size_t>& weights_strides,
+                             std::vector<size_t>& out_dims,
+                             std::vector<size_t>& out_strides)
+{
+    const TensorDescriptor& in = conv_problem.GetIn();
+    std::make_tuple(in_dims, in_strides) = permuteDimsStrides(in.GetLengths(), in.GetStrides());
+
+    const TensorDescriptor& weights = conv_problem.GetWeights();
+    std::make_tuple(weights_dims, weights_strides) =
+        permuteDimsStrides(weights.GetLengths(), weights.GetStrides());
+
+    const TensorDescriptor& out = conv_problem.GetOut();
+    std::make_tuple(out_dims, out_strides) = permuteDimsStrides(out.GetLengths(), out.GetStrides());
+}
+
+MlirConvArgs makeMlirConvArgs(const std::vector<size_t>& in_dims,
+                              const std::vector<size_t>& in_strides,
+                              const std::vector<size_t>& weights_dims,
+                              const std::vector<size_t>& weights_strides,
+                              const std::vector<size_t>& out_dims,
+                              const std::vector<size_t>& out_strides)
+{
+    auto initDimStrides = [](const std::vector<size_t>& dims,
+                             const std::vector<size_t>& strides,
+                             MemRef4DGeneric& target) {
+        std::copy(dims.cbegin(), dims.cend(), &target.sizes[0]);
+        std::copy(strides.cbegin(), strides.cend(), &target.strides[0]);
+    };
+
+    MemRef4DGeneric filter{nullptr, nullptr, 0, {0, 0, 0, 0}, {0, 0, 0, 0}};
+    initDimStrides(weights_dims, weights_strides, filter);
+    MemRef4DGeneric input{nullptr, nullptr, 0, {0, 0, 0, 0}, {0, 0, 0, 0}};
+    initDimStrides(in_dims, in_strides, input);
+    MemRef4DGeneric output{nullptr, nullptr, 0, {0, 0, 0, 0}, {0, 0, 0, 0}};
+    initDimStrides(out_dims, out_strides, output);
+
+    return {filter, input, output};
+}
+} // Anonymous namespace
+
+InvokerFactory MakeMlirFwdInvokerFactory(const ConvolutionContext& ctx)
```

Comment:
My plan for now is to use it in the MLIR bin fwd solver only. I am mostly following `impl_gemm.hpp/cpp` here. Would you rather this to be in the solver? I find encapsulating invoker related stuff in its own component cleaner.

Reply:
[Recommendation]
> Would you rather this to be in the solver? 

Yes, this is correct. 

> I find encapsulating invoker related stuff in its own component cleaner.

Well, it can be considered as a _sub-component_ of this solver. And yes, at first glance, it looks like this makes the code cleaner. But this complicates the code at the same time: When you see a global object you begin asking yourself where it is used. If you see only one use case, you begin asking yourself, why it is used only once. And you decide that there is intent to re-use it somewhere else, later, but this is not true ;)

So I recommend moving this code into the Solver.

[Side note] Defining lambdas in the `GetSolution()` (instead of a function like `MakeInvokerFactory()`) can make the code simpler, because values of all the objects that reside in `GetSolution()` are available for immediate use in the Factory and in the Invoker. Perhaps this is not important for here, just in case.

Reply:
Thanks for the explanation. I'd still prefer invoker to be in its own components because of code re-use possibilities:

I don't plan on this global function to be used in any other solvers. However, between different MLIR solvers, there's a decent chance of code reuse in different direction of invokers: All fwd/bwd/wrw invokers will be using the exact similar format of arguments, regardless of the direction. Besides, the process of deriving the strides and dimensions to pass to MLIR are the same as well in all direction of invokers. 

If I don't put them all in the same compilation unit and instead put them in independent solvers. I will have to copy-paste all above shared components in different solvers. It goes against our original wish to scope code usage, and makes future development and update error-prone as well.

Reply:
> Are you going to use this function in several MLIR Solvers?

Now thinking further on it, my answer would be yes. I will have to come up with new xdlops solver for running on gfx908 and gfx90a machines. This function will be re-used by xdlops fwd solver.

Reply:
Ok